### PR TITLE
feat(wai-handler-launch): Add function to launch arbitrary URLs

### LINKED ISSUE #1: Hard-coded IPv4 address makes IPv6 usage inconvenient
While trying to add [IPv6 support to an application](https://github.com/simonmichael/hledger/issues/1145) that uses `Network.Wai.Handler.Launch.runHostPortUrl`, I noticed that the targeted host is hardcoded to `http://127.0.0.1`. 
This makes adding IPv6 support inconvenient, as it is not possible to open the browser at the correct URL.

The relevant files are:
https://github.com/yesodweb/wai-handlers/blob/master/wai-handler-launch/Network/Wai/Handler/Launch.hs#L182
https://github.com/yesodweb/wai-handlers/blob/master/wai-handler-launch/windows.c#L9

I suspect the easiest way to fix this problem is adding an additional function that also takes a `host` parameter which allows control over the opened URL.

### PR DESCRIPTION
**Pull request by CRTified**
This commit adds the function `runHostPortFullUrl` to allow launching
arbitrary URLs instead of the hardcoded base `http://127.0.0.1`.
It allows using `::1` as the host IP address, while launching `http://[::1]/`,
where the additional brackets are required to access IPv6 IPs from the browser.

As the generation of the default URL is moved into `runHostPortUrl`, both the C launch
function and the Haskell launch function can be reduced to (C)String -> IO ()
without breaking backwards compability.

This closes #1

Before submitting your PR, check that you've:

- [X] Bumped the version number
- [X] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [X] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [ ] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

# Todo

 - [ ] Check for functionality on Windows systems

### CODE REVIEW

**Comment by CRTified** [comment]
Looks like the CI fails are all caused by broken dependencies (QtWebKit) or deprecated functions. I am assuming that they are unrelated to my changes.

File: wai-handler-launch/ChangeLog.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 3.0.2.5
```
Comment (snoyberg) [inline wai-handler-launch/ChangeLog.md]: Since this adds a new function, this should be version 3.0.3

File: wai-handler-launch/windows.c [hunk 2]
```diff
@@ -2,11 +2,7 @@
...
 
-void launch(int port, char *s)
+void launch(char *s)
 {
-    int len = 8 + strlen("http://127.0.0.1:") + strlen(s);
-    char *buff = malloc(len);
-    _snprintf(buff, len, "http://127.0.0.1:%d/%s", port, s);
-    ShellExecute(NULL, "open", buff, NULL, NULL, SW_SHOWNORMAL);
-    free(buff);
+    ShellExecute(NULL, "open", s, NULL, NULL, SW_SHOWNORMAL);
```
Comment (snoyberg) [inline wai-handler-launch/windows.c:7]: Wow, that's so much nicer :)

**Review by snoyberg — CHANGES_REQUESTED** [review]
Minor comment on version numbers. I'll take a crack at getting the CI to pass on `master` as well.

File: wai-handler-launch/ChangeLog.md (hunk 1 above)
Comment (CRTified) [reply wai-handler-launch/ChangeLog.md]: Bumped it accordingly. Somehow forgot to follow semantic versioning. Force-pushed a fix.

**Comment by CRTified** [comment]
Is it preferred to squash the commits together?

File: wai-handler-launch/windows.c (hunk 2 above)
Comment (CRTified) [reply wai-handler-launch/windows.c:7]: I hope that it works. Don't have a windows system for testing in the moment, but I do not see why it shouldn't work.

**Review by snoyberg — APPROVED** [review]
Thanks!

**Comment by simonmichael** [comment]
Unfortunately, this fails to build on Windows, just revealed by hledger's CI: 
https://ci.appveyor.com/project/simonmichael/hledger/builds/30196892

**Comment by simonmichael** [comment]
Ah, I think it should be
```
174 #if WINDOWS
175 launch url = withCString url launch'
```

**Comment by snoyberg** [comment]
Good catch, thanks! I'm uploading the fix shortly.

**Pull request merged**